tests/t0114_gateway_subdomains_test.go

```diff
@@ -2,61 +2,562 @@ package tests
+				)).
+				Response(),
+		))
+
+		with(testGatewayWithManyProtocols(t,
+			"valid parent directory path in directory listing at {cid}.ipfs.example.com/sub/dir",
+			"",
+			Url("%s://%s.ipfs.%s/ipfs/ipns/", u.Scheme, DirCID, u.Host),
+			Expect().
+				Status(200).
+				Body(And(
+					// TODO: implement html expectations
+					Contains("<a href=\"/ipfs/ipns/..\">..</a>"),
+					Contains("<a href=\"/ipfs/ipns/bar\">bar</a>"),
+				)).
+				Response(),
+		))
+
+		with(testGatewayWithManyProtocols(t,
+			"request for deep path resource at {cid}.ipfs.localhost/sub/dir/file",
+			"",
+			Url("%s://%s.ipfs.%s/ipfs/ipns/bar", u.Scheme, DirCID, u.Host),
+			Expect().
+				Status(200).
+				Body(Contains("text-file-content")).
+				Response(),
+		))
+
+		with(testGatewayWithManyProtocols(t,
+			"valid breadcrumb links in the header of directory listing at {cid}.ipfs.example.com/sub/dir",
+			`
+			Note 1: we test for sneaky subdir names  {cid}.ipfs.example.com/ipfs/ipns/ :^)
+			Note 2: example.com/ipfs/.. present in HTML will be redirected to subdomain, so this is expected behavior
+			`,
+			Url("%s://%s.ipfs.%s/ipfs/ipns/", u.Scheme, DirCID, u.Host),
+			Expect().
+				Status(200).
+				Body(
+					And(
+						Contains("Index of"),
+						Contains("/ipfs/<a href=\"//%s/ipfs/%s\">%s</a>/<a href=\"//%s/ipfs/%s/ipfs\">ipfs</a>/<a href=\"//%s/ipfs/%s/ipfs/ipns\">ipns</a>",
+							u.Host, DirCID, DirCID, u.Host, DirCID, u.Host, DirCID),
+					),
+				).
+				Response(),
+		))
+
+		// TODO: # *.ipns.localhost
+		// TODO: # <libp2p-key>.ipns.localhost
+		// TODO: # <dnslink-fqdn>.ipns.localhost
+
+		// # api.localhost/api
+
+		with(testGatewayWithManyProtocols(t,
+			"request for api.localhost returns API response",
+			"Note: we use DIR_CID so refs -r returns some CIDs for child nodes",
+			Url("%s://api.%s/api/v0/refs?arg=%s&r=true", u.Scheme, u.Host, DirCID),
+			Expect().
+				Status(200).
+				Body(Contains("Ref")).
+				Response(),
+		))
+
+		// ## ============================================================================
+		// ## Test DNSLink inlining on HTTP gateways
+		// ## ============================================================================
+
+		// TODO
+
+		// ## ============================================================================
+		// ## Test subdomain-based requests with a custom hostname config
+		// ## (origin per content root at http://*.example.com)
+		// ## ============================================================================
+
+		// # example.com/ip(f|n)s/*
+		// # =============================================================================
+
+		// # path requests to the root hostname should redirect
+		// # to a subdomain URL with proper origin isolation
+
+		// TODO(lidel): for some reason in 114, we only test the simple case, with Host, no proxy or tunnel.
+		// Do we need to port `test_hostname_gateway_response_should_contain` or can we
+		// reuse the same "larger" test?
```

Comment:
I think this could reuse part of the larger test, if we can test localhost and public dns name with the same code, that is better.

I don't remember the reason these have a separate function, but suspect it was to avoid dealing with `--resolve $hostname:127.0.0.1` so we could run tests against a public gateway, but that never happened.